### PR TITLE
[WIP] support block sparse with bf16 and dim=128

### PR DESCRIPTION
**Pull request by kuizhiqing**
I'm trying to extends block sparse with bf16 and d-dim=128.

After a quick test, I've found some thing may not correct in the case dim=128.

Specifically, there is a diff non negligible for q_grad.

```
Forwad diff
Mismatched elements: 1195 / 131072 (0.912%)
Max absolute difference: 0.0004883
Max relative difference: 0.000977

q grad
Mismatched elements: 116949 / 131072 (89.2%)
Max absolute difference: 0.1819
Max relative difference: 844.

k grad
Mismatched elements: 11990 / 131072 (9.15%)
Max absolute difference: 0.0004883
Max relative difference: 0.3

v grad
None
```
Above result was produced by comparer standard flash-attention with sparse flash-attention with block-mask full of 1.

@tridao Can you plz give me some suggestions on this problem, thanks in advance.

### CODE REVIEW

**Comment by kuizhiqing** [comment]
Hi @tridao , can you give me some help with this problem when you available, thank you very much.

One can reproduce my result with script below,
```python
from flash_attn.flash_blocksparse_attn_interface import convert_blockmask
from flash_attn.flash_blocksparse_attn_interface import flash_attn_cuda

import numpy as np
import torch

def run():
    shape = (1, 1024, 1, 128)
    dropout = 0.0
    scale = shape[-1] ** (-0.5)
    is_causal = False

    q = torch.randn(shape, dtype=torch.float16).to(device="cuda:0")
    q.requires_grad = True
    batch_size, seqlen, num_head, head_dim = q.shape
    qt = q.reshape((batch_size * seqlen, q.shape[2], q.shape[3]))
    cu_seqlens = torch.arange(0, (batch_size + 1) * seqlen, step=seqlen, dtype=torch.int32, device=q.device)
    blockmask = torch.full((seqlen//16, seqlen//256), 1, dtype=torch.int32).to(device="cuda:0")
    blockmask = convert_blockmask(blockmask, is_causal)

    out, lse = flash_attn_cuda.fwd_block(qt, qt, qt, cu_seqlens, cu_seqlens, blockmask, seqlen, seqlen, dropout, scale, is_causal, False, None)
    dout = torch.full_like(out, 1)
    dq = torch.empty_like(qt)
    dk = torch.empty_like(qt)
    dv = torch.empty_like(qt)
    flash_attn_cuda.bwd_block(dout, qt, qt, qt, out, lse, dq, dk, dv, cu_seqlens, cu_seqlens, blockmask, seqlen, seqlen, dropout, scale, is_causal, None)

    dout1 = torch.full_like(out, 1)
    out1 = torch.empty_like(out)
    dq1 = torch.empty_like(qt)
    dk1 = torch.empty_like(qt)
    dv1 = torch.empty_like(qt)
    lse1, _ = flash_attn_cuda.fwd(qt, qt, qt, out1, cu_seqlens, cu_seqlens, seqlen, seqlen, dropout, scale, False, False, False, 1, None)
    flash_attn_cuda.bwd(dout1, qt, qt, qt, out1, lse1, dq1, dk1, dv1, cu_seqlens, cu_seqlens, seqlen, seqlen, dropout, scale, False, is_causal, 1, None, None)

    def print_diff(a, b, name="OK"):
        print("--"*10)
        try:
            np.testing.assert_allclose(a, b)
            print(name, "OK")
        except Exception as e:
            print(name, e)

    print_diff(out.cpu(), out1.cpu(), "out")
    print_diff(dq.cpu(), dq1.cpu(), "dq")
    print_diff(dk.cpu(), dk1.cpu(), "dk")
    print_diff(dv.cpu(), dv1.cpu(), "dv")

run()
```
The result may looks like,
```
--------------------
out 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 653 / 131072 (0.498%)
Max absolute difference: 0.001953
Max relative difference: 0.0375
 x: array([[[-0.04132 , -0.74    , -1.903   , ...,  1.183   ,  1.513   ,
         -0.1615  ]],
...
 y: array([[[-0.04132 , -0.74    , -1.903   , ...,  1.183   ,  1.513   ,
         -0.1615  ]],
...
--------------------
dq 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 100766 / 131072 (76.9%)
Max absolute difference: 6.688
Max relative difference: 91.7
 x: array([[[ 8.3876e-04, -4.6158e-03, -1.2978e-02, ...,  1.0178e-02,
          1.2520e-02,  3.6573e-04]],
...
 y: array([[[ 8.378e-04, -4.616e-03, -1.298e-02, ...,  1.018e-02,
          1.252e-02,  3.660e-04]],
...
--------------------
dk 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 61343 / 131072 (46.8%)
Max absolute difference: 0.000557
Max relative difference: 5.45
 x: array([[[-0.0046  , -0.00811 , -0.02122 , ...,  0.009705,  0.012436,
         -0.00582 ]],
...
 y: array([[[-0.0046  , -0.00811 , -0.02122 , ...,  0.009705,  0.012436,
         -0.00582 ]],
...
--------------------
dv 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 128 / 131072 (0.0977%)
Max absolute difference: 0.000977
Max relative difference: 0.000949
 x: array([[[1.023 , 1.023 , 1.023 , ..., 1.023 , 1.023 , 1.023 ]],

       [[1.014 , 1.014 , 1.014 , ..., 1.014 , 1.014 , 1.014 ]],...
 y: array([[[1.023 , 1.023 , 1.023 , ..., 1.023 , 1.023 , 1.023 ]],

       [[1.014 , 1.014 , 1.014 , ..., 1.014 , 1.014 , 1.014 ]],...
```

It seems that there is something wrong with `dq`.

**Comment by tridao** [comment]
Thanks @kuizhiqing. For bf16 I'd expected the numerical error to be larger than fp16 (by the nature of bf16). 
In this simple example, bf16 error is 8 times higher than fp16:
```
In [1]: import torch

In [2]: a = torch.randn(1000, device='cuda', dtype=torch.float16)

In [3]: (a + 0.3 - 0.3 - a).abs().max()
Out[3]: tensor(0.0010, device='cuda:0', dtype=torch.float16)

In [4]: b = torch.randn(1000, device='cuda', dtype=torch.bfloat16)

In [5]: (b + 0.3 - 0.3 - b).abs().max()
Out[5]: tensor(0.0078, device='cuda:0', dtype=torch.bfloat16)
```

I usually compare (FlashAttention in bf16 - standard attention in fp32) vs (standard attention in bf16 - standard attention in fp32).

I'm pretty busy this week (focusing oh PhD thesis) but I'll have more time after Friday to review in details.
Thanks again for your help!
